Implement is_extremal function and tests for issue #16

CreativeBinBag opened the pull request:
## Description

This PR implements the `is_extremal` function, which checks whether a given quantum channel is extremal. The function is based on Section 2.2.4 of Watrous’s *Theory of Quantum Information* and verifies whether the set {A_i† A_j} is linearly independent.

Fixes [#16](https://github.com/vprusso/toqito/issues/16).

## Changes

  -  [x] Implemented the `is_extremal` function in `channel_props/` to determine whether a quantum channel is extremal.
  -  [x] Created `test_is_extremal.py` in `channel_props/tests/` to verify the correctness of `is_extremal`.
  -  [x] Added support for multiple quantum channel representations, including Kraus operators and Choi matrices.
  -  [x] Included test cases for unitary channels, non-extremal channels, Choi matrix input, an example from Watrous's book, and the depolarizing channel.

## Checklist

  -  [x] Used `ruff` for errors related to code style and formatting. All checks were passed.
  -  [x] Verified all previous and newly added unit tests pass in `pytest`.
  -  [x] The Sphinx build completed with 8 warnings, but these are unrelated to this PR. They were present before my changes and occur in auto-generated documentation files.
  -  [x] Used `linkcheck` to check for broken links in the documentation
  -  [x] The Sphinx doctest build completed with 3 failures, but these are unrelated to this PR. They exist in pre-existing autoapi/state_props/ files and seem to be caused by minor floating-point precision differences.

purva-thakre commented on toqito/channel_props/tests/test_is_extremal.py:
```diff
@@ -0,0 +1,61 @@
+"""Tests for the is_extremal function in the toqito library."""
+
+import unittest
```
@CreativeBinBag We use pytest not unittest. Take a look at the existing unit tests to help you.

purva-thakre commented on toqito/channel_props/is_extremal.py:
```diff
@@ -0,0 +1,96 @@
+      - A dictionary with a 'kraus' or 'choi' key,
+      - An object with a callable kraus() method.
+
+    Parameters
```
We use a different style for specifying params etc. Take a look at the contributing guide in the documentation.

CreativeBinBag commented on toqito/channel_props/is_extremal.py:
```diff
@@ -0,0 +1,96 @@
+
+
+def _convert_to_kraus(channel):
+    r"""Convert a quantum channel into its Kraus representation.
```
@purva-thakre the extremality check relies on testing the linear independence of the operators A_i† A_j. The _convert_to_kraus function is defined to normalize the input so that regardless of how the channel is provided, we always have a consistent list of Kraus operators to work with

purva-thakre replied:
I still don't see the need for it. I think there might be confusion between `choi_to_kraus` and ` _convert_to_kraus` if the latter exists in the codebase. 

> we always have a consistent list of Kraus operators to work with

The code for this could be moved into the main function.

purva-thakre commented on toqito/channel_props/is_extremal.py:
```diff
@@ -0,0 +1,96 @@
+            return channel["kraus"]
+        if "choi" in channel:
+            return choi_to_kraus(channel["choi"])
+        raise ValueError("Dictionary channel representation must have either 'kraus' or 'choi' key.")
+    if hasattr(channel, "kraus") and callable(channel.kraus):
+        return channel.kraus()
+    raise ValueError(
+        "Unsupported channel input type. Please provide a list of Kraus operators, "
+        "a Choi matrix, or a dict with a 'kraus' or 'choi' key."
+    )
```
BTW please make sure you add unit tests for these `ValueError`.

purva-thakre commented on toqito/channel_props/is_extremal.py:
```diff
@@ -0,0 +1,96 @@
+    kraus_ops = _convert_to_kraus(channel)
+
+    if not kraus_ops or len(kraus_ops) == 0:
+        raise ValueError("The channel must contain at least one Kraus operator.")
```
Add a unit test for this raised error too.

CreativeBinBag commented:
@purva-thakre I've updated my PR based on your feedback. I integrated the conversion logic into is_extremal.py and added unit tests for invalid inputs. All tests pass locally, but there are docs build issues from pre-existing auto-generated files in autoapi/ which seem unrelated to my changes. Could you please review the updated commit and let me know what else I should change?

vprusso commented:
@CreativeBinBag looks like there are still some failures you may want to look into in the pipeline.

vprusso commented on toqito/channel_props/is_extremal.py:
```diff
@@ -0,0 +1,64 @@
+"""Determines whether a quantum channel is extremal."""
```
You will also want to add this function to the `__init__.py` file in the root of the `channel_props/` folder.

vprusso commented on toqito/channel_props/is_extremal.py:
```diff
@@ -0,0 +1,64 @@
+from toqito.channel_ops.choi_to_kraus import choi_to_kraus
+
+
+def is_extremal(channel):
```
We should assume that `channel` is provided as either a list of Kraus operators or a matrix. We don't want to impose any additional structure on the input that has key values of strings. You can refer to the other functions in the `channel_props/` directory for reference.

vprusso commented on toqito/channel_props/is_extremal.py:
```diff
@@ -0,0 +1,64 @@
+            kraus_ops = choi_to_kraus(channel["choi"])
+        else:
+            raise ValueError("Dictionary must have a 'kraus' or 'choi' key.")
+    elif hasattr(channel, "kraus") and callable(channel.kraus):
+        kraus_ops = channel.kraus()
```
I don't see a scenario in which this would ever work as the input variable won't have this method associated to it.

purva-thakre commented:
@CreativeBinBag Codecov failure is due to some lines in your code not being covered by unit tests. Click on below to see a report on which lines are not covered. 

![image](https://github.com/user-attachments/assets/07f86838-9271-4db5-bca8-10ee09d70363)

purva-thakre commented:
> @purva-thakre I've updated my PR based on your feedback. I integrated the conversion logic into is_extremal.py and added unit tests for invalid inputs. All tests pass locally, but there are docs build issues from pre-existing auto-generated files in autoapi/ which seem unrelated to my changes. Could you please review the updated commit and let me know what else I should change?

I am not too sure about the RTD failure. Let's not worry about that, for now.

purva-thakre commented on toqito/channel_props/tests/test_is_extremal.py:
```diff
@@ -0,0 +1,142 @@
+
+def test_invalid_callable_object():
+    """Ensure an error is raised when an object has a non-callable 'kraus' method."""
+    class InvalidChannel:
```
You could skip invoking this. A variable should be fine to raise the error.

CreativeBinBag commented:
@vprusso @purva-thakre thank you for your feedback! I have tried to follow all your suggestions. One thing to note, I was initially going to use list[np.ndarray] with is_extremal but it seems most functions in the channel_props module accept channels in a nested list format, so I added the nested list format to this function as well. Please let me know what else I should fix

vprusso commented on toqito/channel_props/is_extremal.py:
```diff
@@ -0,0 +1,64 @@
+
+
+def is_extremal(phi: np.ndarray | list[np.ndarray | list[np.ndarray]], tol: float = 1e-9) -> bool:
+    """Determine whether a quantum channel is extremal.
```
We will likely want to add a reference to the textbook by John Watrous (The Theory of Quantum Information). Refer to (for instance) this docstring for how to include citations and include a reference section in the docstring:
https://github.com/vprusso/toqito/blob/f93f5168db7dddd053700b2e92a8badea4509cd9/toqito/channel_props/choi_rank.py#L9

vprusso commented on toqito/channel_props/is_extremal.py:
```diff
@@ -0,0 +1,64 @@
+    the set {A_i^† A_j : i,j=1,...,r} is linearly independent.
+
+    The channel can be provided as either:
+      - A Choi matrix (np.ndarray), which is converted to a flat list of Kraus operators.
```
Since we provide the types in the argument, we don't need the types here in the docstring. This should be for a description about what these inputs pertain to for the purposes of the problem.

vprusso commented on toqito/channel_props/is_extremal.py:
```diff
@@ -0,0 +1,64 @@
+      - A flat list of Kraus operators ([np.ndarray, ...]).
+      - A nested list of Kraus operators ([[np.ndarray, ...], ...]), in which case the
+        list is flattened.
+
```
You may also want to refer to other docstrings in the toqito project (again, using [this](https://github.com/vprusso/toqito/blob/f93f5168db7dddd053700b2e92a8badea4509cd9/toqito/channel_props/choi_rank.py#L9) as an arbitrary example) where we provide sections on examples, references, etc. We would want to include something similar here as well.

vprusso commented on toqito/channel_props/tests/test_is_extremal.py:
```diff
@@ -0,0 +1,132 @@
+@pytest.mark.parametrize(
+    "phi, expected",
+    [
+        # Direct flat list of Kraus operators
+        ([np.array([[0, 1], [1, 0]])], True),
+        # Nested list of Kraus operators
+        ([[np.array([[0, 1], [1, 0]])]], True),
```
```suggestion
        # Direct flat list of Kraus operators.
        ([np.array([[0, 1], [1, 0]])], True),
        # Nested list of Kraus operators.
        ([[np.array([[0, 1], [1, 0]])]], True),
```

vprusso replied:
```suggestion
        # Direct flat list of Kraus operators.
        ([np.array([[0, 1], [1, 0]])], True),
        # Nested list of Kraus operators.
        ([[np.array([[0, 1], [1, 0]])]], True),
```

vprusso commented on toqito/channel_props/tests/test_is_extremal.py:
```diff
@@ -0,0 +1,132 @@
+    assert is_extremal(phi) == expected
+
+
+# Test example from Watrous's book (example 2.33)
+@pytest.mark.parametrize(
+    "phi, expected",
+    [
+        (
+            # Flat list representation.
+            [
+                (1 / np.sqrt(6)) * np.array([[2, 0], [0, 1], [0, 1], [0, 0]]),
+                (1 / np.sqrt(6)) * np.array([[0, 0], [1, 0], [1, 0], [0, 2]]),
+            ],
+            True,
+        ),
+        (
+            # Nested list representation.
+            [
+                [(1 / np.sqrt(6)) * np.array([[2, 0], [0, 1], [0, 1], [0, 0]])],
+                [(1 / np.sqrt(6)) * np.array([[0, 0], [1, 0], [1, 0], [0, 2]])],
+            ],
+            True,
+        ),
+    ],
+)
+def test_example_from_watrous(phi, expected):
+    """Test the example 2.33 from Watrous's *Theory of Quantum Information*."""
+    assert is_extremal(phi) == expected
+
+
+# Test depolarizing channel, which is non-extremal for d > 2 (here d=2)
+@pytest.mark.parametrize(
+    "phi, expected",
+    [
+        (
```
For these two tests (for instance) they can be consolidated into one test that parameterizes over both of the examples. So, the function could be called `test_is_extremal` (for instance), and then you would have test cases in the parameterized section that would pertain to the depolarizing channel test case, the Watrous test case, etc. 

The same would apply to the test cases above this as well, they can likely be parameterized in a similar manner. 

Does that make sense?

vprusso commented on toqito/channel_props/is_extremal.py:
```diff
@@ -0,0 +1,64 @@
+        if len(phi) == 0:
+            raise ValueError("The channel must contain at least one Kraus operator.")
+        if isinstance(phi[0], list):
+            # Flatten the nested list
```
```suggestion
            # Flatten the nested list.
```

vprusso commented on toqito/channel_props/is_extremal.py:
```diff
@@ -0,0 +1,64 @@
+    :type phi: list[numpy.ndarray] | list[list[numpy.ndarray]] | numpy.ndarray
+    :raises ValueError: If the input is neither a valid list of Kraus operators nor a Choi matrix.
+    :return: True if the channel is extremal; False otherwise.
+    :rtype: bool
```
```suggestion
```

vprusso commented on toqito/channel_props/tests/test_is_extremal.py:
```diff
@@ -0,0 +1,132 @@
+
+def test_invalid_list_contents():
+    """Ensure an error is raised when the input list contains invalid elements."""
+    with pytest.raises(ValueError, match="Channel must be a list \\(or nested list\\) of Kraus operators."):
```
I'm surprised you need to escape parens in this way. Could you not do something like
```suggestion
    with pytest.raises(ValueError, match=r"Channel must be a list (or nested list) of Kraus operators."):
```

CreativeBinBag replied:
@vprusso thank you for pointing this out. Even with a raw string, parentheses are treated as capture groups in regex, so they still need to be escaped (\\() to match literally. I could alternatively use re.escape() to handle this automatically. Let me know if you'd prefer that approach

vprusso commented on toqito/channel_props/tests/test_is_extremal.py:
```diff
@@ -0,0 +1,132 @@
+def test_unsupported_input_type():
+    """Ensure an error is raised when the input type is not supported."""
+    with pytest.raises(ValueError, match="Channel must be a list of Kraus operators or a Choi matrix."):
+        is_extremal(42)  # Passing an integer
```
```suggestion
        is_extremal(42)
```

vprusso commented on toqito/channel_props/tests/test_is_extremal.py:
```diff
@@ -0,0 +1,132 @@
+from toqito.channel_props.is_extremal import is_extremal
+
+
+# Test cases for unitary channels (which are extremal)
```
```suggestion
# Test cases for unitary channels (which are extremal).
```

vprusso reviewed (commented):
Looking like it's coming along, @CreativeBinBag ! I left a few comments, but it's getting close!

CreativeBinBag commented:
@vprusso I have tried to incorporate most of the changes you suggested, please have a look! Thank you very much for all the helpful feedback so far!

vprusso reviewed (approved):
Thanks for the alterations, @CreativeBinBag this is great! 

I'll let @purva-thakre give this a look as well, but this looks good to me!

CreativeBinBag commented:
Hi @purva-thakre, could you please let me know if any more changes are needed here?

purva-thakre commented on toqito/channel_props/is_extremal.py:
```diff
@@ -0,0 +1,94 @@
+    (Section 2.2.4: Extremal Channels from :cite:`Watrous_2018_TQI`).
+
+    Theorem 2.31 in :cite:`Watrous_2018_TQI` provides the characterization of extremal
+    quantum channels: A channel :math:`\Phi` is an extreme point of the convex set
```
```suggestion
    quantum channels as a channel :math:`\Phi` is an extreme point of the convex set
```

making a minor change because something is wrong with the workflow

purva-thakre commented on toqito/channel_props/is_extremal.py:
```diff
@@ -0,0 +1,94 @@
+    ...     (1 / np.sqrt(6)) * np.array([[0, 0], [1, 0], [1, 0], [0, 2]])
+    ... ]
+    >>> is_extremal(kraus_ops)
+    True
```
```suggestion
    np.True_
```

purva-thakre commented:
@CreativeBinBag Are you interested in GSoC? Could you please email us so that we can include you in all communications involving GSoC? Our emails are available in the wiki https://github.com/vprusso/toqito/wiki/GSoC-2025-Projects
